### PR TITLE
feat(clerk-js,backend): Throw error if signInUrl is on same origin as a satellite app

### PR DESCRIPTION
**Pull request by desiprisg**
## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [x] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

File: packages/backend/src/tokens/request.ts [hunk 1]
```diff
@@ -102,6 +102,19 @@ function assertProxyUrlOrDomain(proxyUrlOrDomain: string | undefined) {
...
 
+function assertSignInUrlFormatAndOrigin(_signInUrl: string, frontendApi: string) {
+  let signInUrl: URL;
+  try {
+    signInUrl = new URL(_signInUrl);
+  } catch {
+    throw new Error(`The signInUrl needs to have a absolute url format.`);
+  }
+
+  if (signInUrl.hostname === frontendApi.replace('clerk.', '')) {
```
Comment (panteliselef) [inline packages/backend/src/tokens/request.ts:113]: ❓Isn't this handling only the case of a production instance ?

File: packages/clerk-js/src/core/clerk.ts [hunk 2]
```diff
@@ -1280,13 +1282,27 @@ export default class Clerk implements ClerkInterface {
...
+
+    if (this.#options.signInUrl) {
+      let signInUrl: URL;
+      try {
+        signInUrl = new URL(this.#options.signInUrl);
+      } catch {
+        clerkInvalidSignInUrlFormat();
+      }
+
+      if (signInUrl.hostname === this.domain.replace('clerk.', '')) {
```
Comment (panteliselef) [inline packages/clerk-js/src/core/clerk.ts:1302]: @desiprisg Should we update this condition to match the logic from the `backend` package ?

File: packages/clerk-js/src/core/clerk.ts [hunk 3]
```diff
@@ -1280,13 +1282,27 @@ export default class Clerk implements ClerkInterface {
...
+      try {
+        signInUrl = new URL(this.#options.signInUrl);
+      } catch {
+        clerkInvalidSignInUrlFormat();
+      }
+
+      if (signInUrl.origin === window.location.origin) {
+        clerkInvalidSignInUrlOrigin();
+      }
+    }
```
Comment (dimkl) [inline packages/clerk-js/src/core/clerk.ts:1305]: 🙃 
```typescript
  #validateMultiDomainOptions = () => {
   // ...
    if (this.#options.signInUrl) {
      this.#assertSignInUrlFormatAndOrigin(this.#options.signInUrl, window.location.origin);
    }
    // ...
 }
 
#assertSignInUrlFormatAndOrigin = (signInUrl: string, origin: string): never => {
      try {
        signInUrl = new URL(signInUrl);
      } catch {
        clerkInvalidSignInUrlFormat();
      }

      if (signInUrl.origin === window.location.origin) {
        clerkInvalidSignInUrlOrigin();
      }
  }
```

**Review by dimkl — APPROVED** [review]

File: packages/backend/src/tokens/request.test.ts [hunk 4]
```diff
@@ -337,13 +339,15 @@ export default (QUnit: QUnit) => {
...
         isSatellite: true,
+        signInUrl: 'https://primary.dev/sign-in',
         domain: 'satellite.dev',
         userAgent: '[some-agent]',
       });
 
       assertSignedOut(assert, requestState, {
         reason: AuthErrorReason.SatelliteCookieNeedsSyncing,
         isSatellite: true,
+        signInUrl: 'https://primary.dev/sign-in',
```
[the remark this answers is not in the document]
Comment (gkats) [reply packages/backend/src/tokens/request.test.ts:350]: Satellites require `signInUrl` only for development instances.

**Review by panteliselef — APPROVED** [review]

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.

**Pull request merged**